### PR TITLE
Explore: allow users to save Explore state to a new panel in a new dashboard

### LINKED ISSUE #20245: Support an Explore to Dashboard workflow
Add a button in Explore that creates a new dashboard, pre-populated with the current queries being explored.

Often we find interesting things via Explore that we want to build a dashboard around. Right now this means we have to copy/paste the query by hand and create a new dashboard/panel.

This flow works if you navigate from a Dashboard -> Explore -> Dashboard, but not if you start out on a clean-state Explore

### PR DESCRIPTION
**Pull request by Elfo404**
**What this PR does / why we need it**:
Allows Explore users to save the currently explored data tom a new panel in a new dashboard as part of the Explore to Dashboard workflow.

When clicking on the "Add to dashboard" button in the toolbar, the user will be asked to specify in which folder the new dashboard shall be created and with which name.
Screenshot 2022-02-24 at 11 19 32

No further information is to the user, and all queries from the explore pane will be copied over to the existing dashboard (hidden queries will be copied as hidden also in the new panel).

The newly created panel has a default visualization according to the current pane data when available, defaulting to table otherwise. When multiple visualizations are present in the pane (for instance, if different queries produce different visualization types) the first query's preferred visualization type is used).

The modal also allows users to either stay in explore or navigate to the dashboard where the newly created panel has been saved. If the user decides to "keep exploring", a confirmation message is shown informing the user the dashboard has been saved: 
Screenshot 2022-02-24 at 11 20 44

**Which issue(s) this PR fixes**:

Part of #20245 

**Special notes for your reviewer**:

It seems that the error that is supposed to return from the backend when saving a dashboard with the same name as the folder name (https://github.com/grafana/grafana/pull/45148/files#diff-69df9992fce5e8d8e7778774b1ee9dac8d9922fba786d9c8739d5bd760e70af5R71-R74) is never triggered when saving into the default folder (ie. "General" dashboard in the "General" folder). in Dashboards the FE performs this check manually. Shall we consider this a bug in the BE validation?

### CODE REVIEW

File: pkg/models/dashboards.go [hunk 1]
```diff
@@ -70,6 +71,7 @@ var (
 	ErrDashboardWithSameNameAsFolder = DashboardErr{
 		Reason:     "Dashboard name cannot be the same as folder",
 		StatusCode: 400,
+		Status:     "name-match",
```
Comment (Elfo404) [inline pkg/models/dashboards.go:74]: this does not seem to ever trigger

File: public/app/features/explore/Explore.test.tsx [hunk 2]
```diff
@@ -75,6 +75,14 @@ const dummyProps: Props = {
...
       },
     },
+    graphFrames: [],
+    logsFrames: [],
+    tableFrames: [],
+    traceFrames: [],
+    nodeGraphFrames: [],
+    graphResult: null,
+    logsResult: null,
+    tableResult: null,
```
Comment (Elfo404) [inline public/app/features/explore/Explore.test.tsx:85]: this is needed because of the changes in https://github.com/grafana/grafana/pull/45148/files#diff-70e1cc4ad5a5754fe5e6e51db45d1be0501db91555cbfb6db1fc0f80aa328415R200

File: public/app/features/explore/ExploreQueryInspector.test.tsx [hunk 3]
```diff
@@ -38,6 +38,14 @@ const setup = (propOverrides = {}) => {
...
       series: [],
       timeRange: {} as TimeRange,
+      graphFrames: [],
+      logsFrames: [],
+      tableFrames: [],
+      traceFrames: [],
+      nodeGraphFrames: [],
+      graphResult: null,
+      logsResult: null,
+      tableResult: null,
```
Comment (Elfo404) [inline public/app/features/explore/ExploreQueryInspector.test.tsx:48]: this is needed because of the changes in https://github.com/grafana/grafana/pull/45148/files#diff-70e1cc4ad5a5754fe5e6e51db45d1be0501db91555cbfb6db1fc0f80aa328415R200

File: public/app/features/explore/ResponseErrorContainer.test.tsx [hunk 4]
```diff
@@ -50,6 +50,14 @@ function setup(error: DataQueryError) {
...
     state: LoadingState.Error,
     error,
+    graphFrames: [],
+    logsFrames: [],
+    tableFrames: [],
+    traceFrames: [],
+    nodeGraphFrames: [],
+    graphResult: null,
+    logsResult: null,
+    tableResult: null,
```
Comment (Elfo404) [inline public/app/features/explore/ResponseErrorContainer.test.tsx:60]: this is needed because of the changes in https://github.com/grafana/grafana/pull/45148/files#diff-70e1cc4ad5a5754fe5e6e51db45d1be0501db91555cbfb6db1fc0f80aa328415R200

File: public/app/features/explore/state/utils.ts [hunk 5]
```diff
@@ -67,10 +67,18 @@ export const makeExplorePaneState = (): ExploreItemState => ({
...
   series: [],
   timeRange: getDefaultTimeRange(),
+  graphFrames: [],
+  logsFrames: [],
+  traceFrames: [],
+  nodeGraphFrames: [],
+  tableFrames: [],
+  graphResult: null,
+  logsResult: null,
+  tableResult: null,
```
Comment (Elfo404) [inline public/app/features/explore/state/utils.ts:81]: this is needed because of the changes in https://github.com/grafana/grafana/pull/45148/files#diff-70e1cc4ad5a5754fe5e6e51db45d1be0501db91555cbfb6db1fc0f80aa328415R200

File: public/app/types/explore.ts [hunk 6]
```diff
@@ -145,7 +145,7 @@ export interface ExploreItemState {
```
Comment (Elfo404) [inline public/app/types/explore.ts:145]: these are the actual types being used

File: public/app/features/explore/ExploreToolbar.tsx [hunk 7]
```diff
@@ -160,11 +165,47 @@ class UnConnectedExploreToolbar extends PureComponent<Props> {
...
+    if (logsFrames?.some(hasQueryRefId)) {
+      return 'logs';
+    }
+    if (nodeGraphFrames?.some(hasQueryRefId)) {
+      return 'nodeGraph';
+    }
+  }
+
+  return 'table';
+};
```
Comment (Elfo404) [inline public/app/features/explore/ExploreToolbar.tsx:192]: this needs tests

File: public/app/features/explore/AddToDashboard/AddToDashboardModal.tsx [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+import { DataQuery, SelectableValue } from '@grafana/data';
+import { locationService } from '@grafana/runtime';
+import { Button, Field, Input, InputControl, Modal, RadioButtonGroup } from '@grafana/ui';
+import { DashboardPicker } from 'app/core/components/editors/DashboardPicker';
+import { FolderPicker } from 'app/core/components/Select/FolderPicker';
+import { DeepMap, FieldError, useForm } from 'react-hook-form';
+import { addToDashboard, SaveToExistingDashboardDTO, SaveToNewDashboardDTO } from './addToDashboard';
+// import { notifyApp } from 'app/core/actions';
+// import { createErrorNotification, createSuccessNotification } from 'app/core/copy/appNotification';
+// import { useDispatch } from 'react-redux';
```
Comment (Elfo404) [inline public/app/features/explore/AddToDashboard/AddToDashboardModal.tsx:11]: this (and related comments in this file) is still WIP, need to check with UX the best way to provide users feedback upon success/failure

File: public/app/features/explore/AddToDashboard/addToDashboard.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  visualization: string;
+}
+
+export interface SaveToExistingDashboardDTO {
+  dashboard: string;
+  queries: DataQuery[];
+  visualization: string;
+}
+
+const createDashboardApiCall = (dashboard: DashboardDataDTO, folderId: number) => {
```
Comment (Elfo404) [inline public/app/features/explore/AddToDashboard/addToDashboard.ts:20]: still WIP, i'd like to move this to `dashboardSrv`, need to check with user essentials

File: public/app/features/explore/AddToDashboard/addToDashboard.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+const createDashboard = (dashboardName: string, folderId: number, queries: DataQuery[], visualization: string) => {
+  const dashboard = getDashboardSrv().create({ title: dashboardName }, { folderId });
+
+  dashboard.addPanel({ targets: queries, type: visualization });
+
+  return lastValueFrom(createDashboardApiCall(dashboard.getSaveModelClone(), folderId));
+};
+
+export const addToDashboard = (data: SaveToNewDashboardDTO | SaveToExistingDashboardDTO) => {
+  // TODO: if create dashboard...
```
Comment (Elfo404) [inline public/app/features/explore/AddToDashboard/addToDashboard.ts:42]: i'm going to remove this together with everything related to `SaveToExistingDashboardDTO` to keep PRs smaller.

File: public/app/features/explore/ExploreToolbar.tsx (hunk 7 above)
Comment (Elfo404) [reply public/app/features/explore/ExploreToolbar.tsx]: this has tests now

File: public/app/features/explore/AddToDashboard/AddToDashboardModal.tsx [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+    if (error) {
+      switch (error.status) {
+        case 'name-exists':
+        case 'empty-name':
+        case 'name-match':
+          setError('dashboardName', { message: error.message, shouldFocus: true });
+          break;
+        default:
+        // TODO: Other unknown errors may happen, we should handle them by displaying an error message
```
Comment (Elfo404) [inline public/app/features/explore/AddToDashboard/AddToDashboardModal.tsx:55]: Needs UX: how shall we show errors that are not related to a specific field?

An alert notification (the one appearing on the top-right corner) are not a viable solution as the modal (for accessibility reasons) is going to be the only component readable by screen readers, so the error must be in the modal itself.

cc. @nadinevehling

File: pkg/models/dashboards.go (hunk 1 above)
Comment (Elfo404) [reply pkg/models/dashboards.go:74]: FYI @ifrost if you want to take a look tomorrow, this seems to trigger only when NOT saving a dashboard in the "General" folder, for each other folder it seems to trigger correctly

Comment (ifrost) [reply pkg/models/dashboards.go:74]: For the record: the backend team is having a discussion about it what should actually change.

File: public/app/features/explore/AddToDashboard/__test__/utils.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  items: [],
+  url: '',
+  uri: '',
+  tags: [],
+});
+
+const foldersSearchPromise = Promise.resolve([createFolder('Folder 1', 0), createFolder('Folder 2', 0)]);
+jest.spyOn(dashboardApi, 'searchFolders').mockReturnValue(foldersSearchPromise);
+
+export const openModal = async () => {
```
Comment (ifrost) [inline public/app/features/explore/AddToDashboard/__test__/utils.ts:20]: Is it being used anywhere? I don't see any direct imports 🤔

File: public/app/features/explore/AddToDashboard/__test__/utils.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+  userEvent.click(screen.getByRole('button', { name: /add to dashboard/i }));
+
+  return act(async () => {
+    // FolderPicker asyncrounously sets its internal state based on search results, causing ugly warnings when testing.
```
Comment (ifrost) [inline public/app/features/explore/AddToDashboard/__test__/utils.ts:24]:
```suggestion
    // FolderPicker asynchronously sets its internal state based on search results, causing ugly warnings when testing.
```

File: public/app/features/explore/AddToDashboard/AddToDashboardModal.test.tsx [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+      });
+
+      expect(saveMock).toHaveBeenCalledWith(
+        {
+          dashboardName: dashboardNameInput.value,
+          queries: [],
+          visualization: 'table',
+          folderId: 1,
+        },
+        expect.anything()
```
Comment (ifrost) [inline public/app/features/explore/AddToDashboard/AddToDashboardModal.test.tsx:84]: Would it make sense to assert the redirect as well?

```suggestion
        false
```

File: public/app/features/explore/AddToDashboard/index.test.tsx [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+    expect(addToDashboardMock).toHaveBeenCalledWith(
+      expect.objectContaining({
+        queries: queries,
+        visualization: 'timeseries',
+      })
+    );
+  });
+
+  it('Sets visualization to logs if there are log frames', async () => {
+    const queries: DataQuery[] = [{ refId: 'A' }];
```
Comment (ifrost) [inline public/app/features/explore/AddToDashboard/index.test.tsx:182]: There's some code duplication in this test and tests below (first 11 lines). Could it be extracted to a helper function?

File: public/app/features/explore/AddToDashboard/__test__/utils.ts (hunk 12 above)
Comment (Elfo404) [reply public/app/features/explore/AddToDashboard/__test__/utils.ts]: good point, it's a leftover file from previous tests, remove in 9d481c1449

File: public/app/features/explore/AddToDashboard/__test__/utils.ts (hunk 13 above)
Comment (Elfo404) [reply public/app/features/explore/AddToDashboard/__test__/utils.ts]: deleted :)

File: public/app/features/explore/AddToDashboard/AddToDashboardModal.test.tsx (hunk 14 above)
Comment (Elfo404) [reply public/app/features/explore/AddToDashboard/AddToDashboardModal.test.tsx:84]: here i'm only testing if the form submits correctly, given  the navigation logic it's not handled in the tested component but in the wrapper one (`index.tesx`) is tested in https://github.com/grafana/grafana/blob/909130cde9bef0e9f937b306cbe12675fe1bdcdd/public/app/features/explore/AddToDashboard/index.test.tsx#L75

File: public/app/features/explore/AddToDashboard/index.test.tsx (hunk 15 above)
Comment (Elfo404) [reply public/app/features/explore/AddToDashboard/index.test.tsx:182]: i'm not sure about this, the setup function already has some unopinionated default values (no queries & no response). each test is actually "duplicating" this bit by passing in different queries to the setup function, it's just that for this and the following ones they happen to be the same. how would you approach it?

**Review by gelicia — CHANGES_REQUESTED** [review]
Just a couple small things! Looks great!

File: public/app/features/explore/AddToDashboard/AddToDashboardModal.tsx [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+    if (error) {
+      switch (error.status) {
+        case 'name-exists':
+        case 'empty-name':
+        case 'name-match':
+          // error.message should always be defined here
+          setError('dashboardName', { message: error.message ?? 'This field is invalid' });
+          break;
+        default:
+          setSubmissionError(
```
[the remark this answers is not in the document]
Comment (Elfo404) [reply public/app/features/explore/AddToDashboard/AddToDashboardModal.tsx:49]: we could and it was actually my first approach but it comes with a few caveats:
- setError is designed to be field-related, there's an feature request for supporting this use case at https://github.com/react-hook-form/react-hook-form/issues/7661
- there's some extra logic to handle to clean the errors from the form as manually set errors that are not bound to input fields are not automatically removed when resubmitting. so it needs to be done manually

i've implemented it in https://github.com/grafana/grafana/pull/45995 if you want to take a look, i have no strong preference between this solution or the other as both of them come with their own pros and cons

[the remark this answers is not in the document]
Comment (gelicia) [reply public/app/features/explore/AddToDashboard/AddToDashboardModal.tsx:49]: Ah sorry, it does make sense to me to keep what you have when you explain the constraint in react-hook-form. The form takes care of errors on individual fields and your component takes care of errors that might not be tied to a specific field.

**Review by gelicia — APPROVED** [review]

**Review by ifrost — APPROVED** [review]

File: public/app/features/explore/AddToDashboard/index.test.tsx (hunk 15 above)
Comment (ifrost) [reply public/app/features/explore/AddToDashboard/index.test.tsx:182]: Probably just a helper function that would run common steps and allow some customization with params. Nothing major, it can stay as it is.
